// @syncfusion/ej2-angular-calendars/esm2020/src/timepicker/timepicker-all.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { TimePickerModule } from './timepicker.module';
import * as i0 from "@angular/core";
/**
 * NgModule definition for the TimePicker component with providers.
 */
export class TimePickerAllModule {
}
TimePickerAllModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.0.3", ngImport: i0, type: TimePickerAllModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TimePickerAllModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.0.3", ngImport: i0, type: TimePickerAllModule, imports: [CommonModule, TimePickerModule], exports: [TimePickerModule] });
TimePickerAllModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.0.3", ngImport: i0, type: TimePickerAllModule, providers: [], imports: [[CommonModule, TimePickerModule], TimePickerModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.0.3", ngImport: i0, type: TimePickerAllModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, TimePickerModule],
                    exports: [
                        TimePickerModule
                    ],
                    providers: []
                }]
        }] });
