// @syncfusion/ej2-angular-dropdowns/esm2020/src/drop-down-tree/dropdowntree.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { DropDownTreeComponent } from './dropdowntree.component';
import * as i0 from "@angular/core";
/**
 * NgModule definition for the DropDownTree component.
 */
export class DropDownTreeModule {
}
DropDownTreeModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.0.3", ngImport: i0, type: DropDownTreeModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
DropDownTreeModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.0.3", ngImport: i0, type: DropDownTreeModule, declarations: [DropDownTreeComponent], imports: [CommonModule], exports: [DropDownTreeComponent] });
DropDownTreeModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.0.3", ngImport: i0, type: DropDownTreeModule, imports: [[CommonModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.0.3", ngImport: i0, type: DropDownTreeModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule],
                    declarations: [
                        DropDownTreeComponent
                    ],
                    exports: [
                        DropDownTreeComponent
                    ]
                }]
        }] });
