// @syncfusion/ej2-angular-dropdowns/esm2020/src/drop-down-tree/dropdowntree-all.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { DropDownTreeModule } from './dropdowntree.module';
import * as i0 from "@angular/core";
/**
 * NgModule definition for the DropDownTree component with providers.
 */
export class DropDownTreeAllModule {
}
DropDownTreeAllModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.0.3", ngImport: i0, type: DropDownTreeAllModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
DropDownTreeAllModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.0.3", ngImport: i0, type: DropDownTreeAllModule, imports: [CommonModule, DropDownTreeModule], exports: [DropDownTreeModule] });
DropDownTreeAllModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.0.3", ngImport: i0, type: DropDownTreeAllModule, providers: [], imports: [[CommonModule, DropDownTreeModule], DropDownTreeModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.0.3", ngImport: i0, type: DropDownTreeAllModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, DropDownTreeModule],
                    exports: [
                        DropDownTreeModule
                    ],
                    providers: []
                }]
        }] });
