// @syncfusion/ej2-angular-dropdowns/esm2020/src/auto-complete/autocomplete-all.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { AutoCompleteModule } from './autocomplete.module';
import * as i0 from "@angular/core";
/**
 * NgModule definition for the AutoComplete component with providers.
 */
export class AutoCompleteAllModule {
}
AutoCompleteAllModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.0.3", ngImport: i0, type: AutoCompleteAllModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
AutoCompleteAllModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.0.3", ngImport: i0, type: AutoCompleteAllModule, imports: [CommonModule, AutoCompleteModule], exports: [AutoCompleteModule] });
AutoCompleteAllModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.0.3", ngImport: i0, type: AutoCompleteAllModule, providers: [], imports: [[CommonModule, AutoCompleteModule], AutoCompleteModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.0.3", ngImport: i0, type: AutoCompleteAllModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, AutoCompleteModule],
                    exports: [
                        AutoCompleteModule
                    ],
                    providers: []
                }]
        }] });
