// @syncfusion/ej2-angular-calendars/esm2020/src/calendar/calendar-all.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { CalendarModule } from './calendar.module';
import { Islamic } from '@syncfusion/ej2-calendars';
import * as i0 from "@angular/core";
export const IslamicService = { provide: 'CalendarsIslamic', useValue: Islamic };
/**
 * NgModule definition for the Calendar component with providers.
 */
export class CalendarAllModule {
}
CalendarAllModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.0.3", ngImport: i0, type: CalendarAllModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
CalendarAllModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.0.3", ngImport: i0, type: CalendarAllModule, imports: [CommonModule, CalendarModule], exports: [CalendarModule] });
CalendarAllModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.0.3", ngImport: i0, type: CalendarAllModule, providers: [
        IslamicService
    ], imports: [[CommonModule, CalendarModule], CalendarModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.0.3", ngImport: i0, type: CalendarAllModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, CalendarModule],
                    exports: [
                        CalendarModule
                    ],
                    providers: [
                        IslamicService
                    ]
                }]
        }] });
