// @syncfusion/ej2-angular-dropdowns/esm2020/src/mention/mention.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { MentionComponent } from './mention.component';
import * as i0 from "@angular/core";
/**
 * NgModule definition for the Mention component.
 */
export class MentionModule {
}
MentionModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.0.3", ngImport: i0, type: MentionModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
MentionModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.0.3", ngImport: i0, type: MentionModule, declarations: [MentionComponent], imports: [CommonModule], exports: [MentionComponent] });
MentionModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.0.3", ngImport: i0, type: MentionModule, imports: [[CommonModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.0.3", ngImport: i0, type: MentionModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule],
                    declarations: [
                        MentionComponent
                    ],
                    exports: [
                        MentionComponent
                    ]
                }]
        }] });
