// @syncfusion/ej2-angular-dropdowns/esm2020/src/mention/mention-all.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { MentionModule } from './mention.module';
import * as i0 from "@angular/core";
/**
 * NgModule definition for the Mention component with providers.
 */
export class MentionAllModule {
}
MentionAllModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.0.3", ngImport: i0, type: MentionAllModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
MentionAllModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.0.3", ngImport: i0, type: MentionAllModule, imports: [CommonModule, MentionModule], exports: [MentionModule] });
MentionAllModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.0.3", ngImport: i0, type: MentionAllModule, providers: [], imports: [[CommonModule, MentionModule], MentionModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.0.3", ngImport: i0, type: MentionAllModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, MentionModule],
                    exports: [
                        MentionModule
                    ],
                    providers: []
                }]
        }] });
