// @syncfusion/ej2-angular-schedule/esm2020/src/recurrence-editor/recurrenceeditor.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { RecurrenceEditorComponent } from './recurrenceeditor.component';
import * as i0 from "@angular/core";
/**
 * NgModule definition for the RecurrenceEditor component.
 */
export class RecurrenceEditorModule {
}
RecurrenceEditorModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.0.3", ngImport: i0, type: RecurrenceEditorModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
RecurrenceEditorModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.0.3", ngImport: i0, type: RecurrenceEditorModule, declarations: [RecurrenceEditorComponent], imports: [CommonModule], exports: [RecurrenceEditorComponent] });
RecurrenceEditorModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.0.3", ngImport: i0, type: RecurrenceEditorModule, imports: [[CommonModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.0.3", ngImport: i0, type: RecurrenceEditorModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule],
                    declarations: [
                        RecurrenceEditorComponent
                    ],
                    exports: [
                        RecurrenceEditorComponent
                    ]
                }]
        }] });
