// @syncfusion/ej2-angular-schedule/esm2020/src/recurrence-editor/recurrenceeditor-all.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { RecurrenceEditorModule } from './recurrenceeditor.module';
import * as i0 from "@angular/core";
/**
 * NgModule definition for the RecurrenceEditor component with providers.
 */
export class RecurrenceEditorAllModule {
}
RecurrenceEditorAllModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.0.3", ngImport: i0, type: RecurrenceEditorAllModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
RecurrenceEditorAllModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.0.3", ngImport: i0, type: RecurrenceEditorAllModule, imports: [CommonModule, RecurrenceEditorModule], exports: [RecurrenceEditorModule] });
RecurrenceEditorAllModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.0.3", ngImport: i0, type: RecurrenceEditorAllModule, providers: [], imports: [[CommonModule, RecurrenceEditorModule], RecurrenceEditorModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.0.3", ngImport: i0, type: RecurrenceEditorAllModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, RecurrenceEditorModule],
                    exports: [
                        RecurrenceEditorModule
                    ],
                    providers: []
                }]
        }] });
