// @syncfusion/ej2-angular-schedule/esm2020/src/schedule/schedule.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ViewDirective, ViewsDirective } from './views.directive';
import { ResourceDirective, ResourcesDirective } from './resources.directive';
import { HeaderRowDirective, HeaderRowsDirective } from './headerrows.directive';
import { ScheduleComponent } from './schedule.component';
import * as i0 from "@angular/core";
/**
 * NgModule definition for the Schedule component.
 */
export class ScheduleModule {
}
ScheduleModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.0.3", ngImport: i0, type: ScheduleModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
ScheduleModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.0.3", ngImport: i0, type: ScheduleModule, declarations: [ScheduleComponent,
        ViewDirective,
        ViewsDirective,
        ResourceDirective,
        ResourcesDirective,
        HeaderRowDirective,
        HeaderRowsDirective], imports: [CommonModule], exports: [ScheduleComponent,
        ViewDirective,
        ViewsDirective,
        ResourceDirective,
        ResourcesDirective,
        HeaderRowDirective,
        HeaderRowsDirective] });
ScheduleModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.0.3", ngImport: i0, type: ScheduleModule, imports: [[CommonModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.0.3", ngImport: i0, type: ScheduleModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule],
                    declarations: [
                        ScheduleComponent,
                        ViewDirective,
                        ViewsDirective,
                        ResourceDirective,
                        ResourcesDirective,
                        HeaderRowDirective,
                        HeaderRowsDirective
                    ],
                    exports: [
                        ScheduleComponent,
                        ViewDirective,
                        ViewsDirective,
                        ResourceDirective,
                        ResourcesDirective,
                        HeaderRowDirective,
                        HeaderRowsDirective
                    ]
                }]
        }] });
