// @syncfusion/ej2-angular-schedule/esm2020/src/schedule/schedule-all.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ScheduleModule } from './schedule.module';
import { Day, Week, WorkWeek, Month, Year, Agenda, MonthAgenda, TimelineViews, TimelineMonth, TimelineYear, Resize, DragAndDrop, ExcelExport, ICalendarExport, ICalendarImport, Print } from '@syncfusion/ej2-schedule';
import * as i0 from "@angular/core";
export const DayService = { provide: 'ScheduleDay', useValue: Day };
export const WeekService = { provide: 'ScheduleWeek', useValue: Week };
export const WorkWeekService = { provide: 'ScheduleWorkWeek', useValue: WorkWeek };
export const MonthService = { provide: 'ScheduleMonth', useValue: Month };
export const YearService = { provide: 'ScheduleYear', useValue: Year };
export const AgendaService = { provide: 'ScheduleAgenda', useValue: Agenda };
export const MonthAgendaService = { provide: 'ScheduleMonthAgenda', useValue: MonthAgenda };
export const TimelineViewsService = { provide: 'ScheduleTimelineViews', useValue: TimelineViews };
export const TimelineMonthService = { provide: 'ScheduleTimelineMonth', useValue: TimelineMonth };
export const TimelineYearService = { provide: 'ScheduleTimelineYear', useValue: TimelineYear };
export const ResizeService = { provide: 'ScheduleResize', useValue: Resize };
export const DragAndDropService = { provide: 'ScheduleDragAndDrop', useValue: DragAndDrop };
export const ExcelExportService = { provide: 'ScheduleExcelExport', useValue: ExcelExport };
export const ICalendarExportService = { provide: 'ScheduleICalendarExport', useValue: ICalendarExport };
export const ICalendarImportService = { provide: 'ScheduleICalendarImport', useValue: ICalendarImport };
export const PrintService = { provide: 'SchedulePrint', useValue: Print };
/**
 * NgModule definition for the Schedule component with providers.
 */
export class ScheduleAllModule {
}
ScheduleAllModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.0.3", ngImport: i0, type: ScheduleAllModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
ScheduleAllModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.0.3", ngImport: i0, type: ScheduleAllModule, imports: [CommonModule, ScheduleModule], exports: [ScheduleModule] });
ScheduleAllModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.0.3", ngImport: i0, type: ScheduleAllModule, providers: [
        DayService,
        WeekService,
        WorkWeekService,
        MonthService,
        YearService,
        AgendaService,
        MonthAgendaService,
        TimelineViewsService,
        TimelineMonthService,
        TimelineYearService,
        ResizeService,
        DragAndDropService,
        ExcelExportService,
        ICalendarExportService,
        ICalendarImportService,
        PrintService
    ], imports: [[CommonModule, ScheduleModule], ScheduleModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.0.3", ngImport: i0, type: ScheduleAllModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, ScheduleModule],
                    exports: [
                        ScheduleModule
                    ],
                    providers: [
                        DayService,
                        WeekService,
                        WorkWeekService,
                        MonthService,
                        YearService,
                        AgendaService,
                        MonthAgendaService,
                        TimelineViewsService,
                        TimelineMonthService,
                        TimelineYearService,
                        ResizeService,
                        DragAndDropService,
                        ExcelExportService,
                        ICalendarExportService,
                        ICalendarImportService,
                        PrintService
                    ]
                }]
        }] });
