// @syncfusion/ej2-angular-calendars/esm2020/src/daterangepicker/daterangepicker-all.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { DateRangePickerModule } from './daterangepicker.module';
import * as i0 from "@angular/core";
/**
 * NgModule definition for the DateRangePicker component with providers.
 */
export class DateRangePickerAllModule {
}
DateRangePickerAllModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.0.3", ngImport: i0, type: DateRangePickerAllModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
DateRangePickerAllModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.0.3", ngImport: i0, type: DateRangePickerAllModule, imports: [CommonModule, DateRangePickerModule], exports: [DateRangePickerModule] });
DateRangePickerAllModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.0.3", ngImport: i0, type: DateRangePickerAllModule, providers: [], imports: [[CommonModule, DateRangePickerModule], DateRangePickerModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.0.3", ngImport: i0, type: DateRangePickerAllModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, DateRangePickerModule],
                    exports: [
                        DateRangePickerModule
                    ],
                    providers: []
                }]
        }] });
