// @syncfusion/ej2-angular-calendars/esm2020/src/daterangepicker/daterangepicker.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { PresetDirective, PresetsDirective } from './presets.directive';
import { DateRangePickerComponent } from './daterangepicker.component';
import * as i0 from "@angular/core";
/**
 * NgModule definition for the DateRangePicker component.
 */
export class DateRangePickerModule {
}
DateRangePickerModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.0.3", ngImport: i0, type: DateRangePickerModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
DateRangePickerModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.0.3", ngImport: i0, type: DateRangePickerModule, declarations: [DateRangePickerComponent,
        PresetDirective,
        PresetsDirective], imports: [CommonModule], exports: [DateRangePickerComponent,
        PresetDirective,
        PresetsDirective] });
DateRangePickerModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.0.3", ngImport: i0, type: DateRangePickerModule, imports: [[CommonModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.0.3", ngImport: i0, type: DateRangePickerModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule],
                    declarations: [
                        DateRangePickerComponent,
                        PresetDirective,
                        PresetsDirective
                    ],
                    exports: [
                        DateRangePickerComponent,
                        PresetDirective,
                        PresetsDirective
                    ]
                }]
        }] });
