// @syncfusion/ej2-angular-calendars/esm2020/src/datepicker/datepicker-all.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { DatePickerModule } from './datepicker.module';
import { MaskedDateTime } from '@syncfusion/ej2-calendars';
import * as i0 from "@angular/core";
export const MaskedDateTimeService = { provide: 'CalendarsMaskedDateTime', useValue: MaskedDateTime };
/**
 * NgModule definition for the DatePicker component with providers.
 */
export class DatePickerAllModule {
}
DatePickerAllModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.0.3", ngImport: i0, type: DatePickerAllModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
DatePickerAllModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.0.3", ngImport: i0, type: DatePickerAllModule, imports: [CommonModule, DatePickerModule], exports: [DatePickerModule] });
DatePickerAllModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.0.3", ngImport: i0, type: DatePickerAllModule, providers: [
        MaskedDateTimeService
    ], imports: [[CommonModule, DatePickerModule], DatePickerModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.0.3", ngImport: i0, type: DatePickerAllModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, DatePickerModule],
                    exports: [
                        DatePickerModule
                    ],
                    providers: [
                        MaskedDateTimeService
                    ]
                }]
        }] });
