// @syncfusion/ej2-angular-calendars/esm2020/src/datetimepicker/datetimepicker-all.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { DateTimePickerModule } from './datetimepicker.module';
import * as i0 from "@angular/core";
/**
 * NgModule definition for the DateTimePicker component with providers.
 */
export class DateTimePickerAllModule {
}
DateTimePickerAllModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.0.3", ngImport: i0, type: DateTimePickerAllModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
DateTimePickerAllModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.0.3", ngImport: i0, type: DateTimePickerAllModule, imports: [CommonModule, DateTimePickerModule], exports: [DateTimePickerModule] });
DateTimePickerAllModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.0.3", ngImport: i0, type: DateTimePickerAllModule, providers: [], imports: [[CommonModule, DateTimePickerModule], DateTimePickerModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.0.3", ngImport: i0, type: DateTimePickerAllModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, DateTimePickerModule],
                    exports: [
                        DateTimePickerModule
                    ],
                    providers: []
                }]
        }] });
